{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6d2da9ef-1ced-4fd1-96ee-beac96617c3e",
   "metadata": {},
   "source": [
    "# Step 5 -- Genome Mapping\n",
    "-- Alex Warwick Vesztrocy, February 2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "072f5604-51f6-41fa-bd06-32ed24212615",
   "metadata": {},
   "source": [
    "Generate genome mapping tables for each of the species. This is necessary as PANTHER genomes are imported from UniProt RPs, whereas bgee is using ensembl data directly."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f56ba682-c169-4089-b9f8-eb41372918bf",
   "metadata": {},
   "source": [
    "This is going to be a combination of 3 mapping techniques for most of the files.\n",
    "\n",
    "    1. PANTHER cross references inside the tree files\n",
    "    2. bgee-uniprot cross references from the UniProt website\n",
    "    3. ensembl-uniprot cross references taken from the relevant assembly used in bgee\n",
    "\n",
    "Then for certain species these will be supplemented with bespoke mapping files from 3rd party resources. e.g., RAT, MOUSE, HUMAN.\n",
    "\n",
    "    - RAT: using RGD gene definition file (contains ensembl cross-references). Downloaded on 7.3.24\n",
    "    - MOUSE: using MGI gene model co-ordinates file (contains ensebl gene IDs). Downloaded on 7.3.24\n",
    "    - HUMAN: use HGNC to ensembl mapping. Generated and downloaded on 7.3.24\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a9e6ab5-77ed-4ae4-b434-e44e4af91316",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a7e80f10-3003-4c74-bfe3-aedbe5d64c37",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm.auto import tqdm\n",
    "import pandas as pd\n",
    "\n",
    "from lib.PantherParser import PantherTrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1f6451df-b662-46d0-b8d1-2e56ceff761f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt = PantherTrees(\n",
    "     \"./data/panther-18.0/trees/\",\n",
    "     taxonomy_fn=\"./data/panther-18.0/species_tree.nhx\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "80ff8bd7-37dd-4fc2-a27b-5a19de34f610",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_panther_mapping(pt):\n",
    "    def load():\n",
    "        for fam in pt.iter_fams():\n",
    "            for g in fam.genes.values():\n",
    "                r = {'fam_id': fam.id} | g\n",
    "                for (k, v) in r.items():\n",
    "                    if k.lower().startswith('ensembl'):\n",
    "                        r[k] = v.split('.')[0]\n",
    "                yield r\n",
    "    return pd.DataFrame(load())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4fd0a3e3-d17b-4d8d-b0e5-8fe4813d7982",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "42d11dd0475e4473bba41980d3d1d3ad",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "For each family:   0%|          | 0/15693 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pthr_genes = load_panther_mapping(pt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fe61ca3f-f1ec-4d17-8516-4ba74ab30150",
   "metadata": {},
   "outputs": [],
   "source": [
    "# species that have data and are in PANTHER\n",
    "dataset = {'animals': {'ANOCA','BOVIN','CAEEL','CANLF','CHICK','DANRE','DROME','FELCA','GORGO','HORSE','HUMAN','LEPOC','MACMU','MONDO','MOUSE','ORNAN','ORYLA','PANTR','PIG','RAT', 'XENTR'},\n",
    "           'plants': {'AMBTC','ARATH','BRANA','CAPAN','CUCSA','HELAN','MAIZE','MANES','MARPO','MEDTR','ORYSJ','PHYPA','POPTR','SELML','SETIT','SOLLC','SOLTU','SOYBN','VITVI','WHEAT'}}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d9b47a9-946e-4cc7-ae9f-b8f262f905fb",
   "metadata": {},
   "source": [
    "---\n",
    "##Â Animal Gene Mapping to bgee\n",
    "\n",
    "Ensuring we get a good mapping between bgee identifiers and PANTHER."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "05a57ad2-c8ce-4808-b0cc-532cf33ba425",
   "metadata": {},
   "outputs": [],
   "source": [
    "pthr_genes_animal = pthr_genes[pthr_genes.species.isin(dataset['animals'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "75d20e71-87fd-4eab-8db4-5267dcfb18fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the UniProt - bgee mapping file\n",
    "bgee_map = pd.read_csv('./data/uniprot/bgee-uniprot.tsv.gz', sep='\\t')\n",
    "# remove the trailing ';'\n",
    "bgee_map.Bgee = bgee_map.Bgee.apply(lambda x: x[:-1])\n",
    "bgee_map.rename(columns={'Entry': 'UniProtKB'}, inplace=True)\n",
    "bgee_map = bgee_map[['UniProtKB', 'Bgee']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d19bfe4a-1d5f-4493-9587-c3fa601afd85",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load HGNC map\n",
    "hgnc_map = pd.read_csv('./data/mapping/hgnc/hgnc.map', sep='\\t')\n",
    "hgnc_map.rename(columns={hgnc_map.keys()[0]: 'hgnc_id',\n",
    "                         hgnc_map.keys()[1]: 'ensembl_hgnc_map'},\n",
    "                inplace=True)\n",
    "hgnc_map.hgnc_id = hgnc_map.hgnc_id.apply(lambda x: x.split(':')[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d30c093b-2e08-47bc-b2fd-77a84fcdae65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load RGD map\n",
    "def load_rgd():\n",
    "    def gen():\n",
    "        rgd_map = pd.read_csv('./data/mapping/rgd/GENES_RAT.txt', sep='\\t', comment='#', low_memory=False)\n",
    "        f = rgd_map.ENSEMBL_ID.notna()\n",
    "        for (_, r) in rgd_map[f][['GENE_RGD_ID', 'ENSEMBL_ID']].iterrows():\n",
    "            for x in r['ENSEMBL_ID'].split(';'):\n",
    "                yield (str(r['GENE_RGD_ID']), x)\n",
    "    return pd.DataFrame(gen(), columns=['rgd_id', 'ensembl_rgd_map'])\n",
    "\n",
    "rgd_map = load_rgd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6d6533a4-90e3-49f4-b16c-beaef72a7ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load MGI map\n",
    "def load_mgi():\n",
    "    mgi_map = pd.read_csv('./data/mapping/mgi/MGI_Gene_Model_Coord.rpt', sep='\\t', low_memory=False, index_col=False)\n",
    "    mgi_map = mgi_map.rename(columns={'1. MGI accession id': 'mgi_id',\n",
    "                                      '11. Ensembl gene id': 'ensembl_mgi_map'})\n",
    "    mgi_map = mgi_map[mgi_map['ensembl_mgi_map'].notna()][['mgi_id', 'ensembl_mgi_map']]\n",
    "    mgi_map['mgi_id'] = mgi_map['mgi_id'].apply(lambda x: 'MGI={}'.format(x.split(':')[1]))\n",
    "    return mgi_map\n",
    "\n",
    "mgi_map = load_mgi()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b63234df-5138-46fc-ba7e-bfb5e33e3c5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "141c061bc06d42c3bfcebc1ad9495cbd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/21 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stats = []\n",
    "res = {}\n",
    "for (sp, zdf) in tqdm(pthr_genes_animal.groupby('species')):\n",
    "    # join with the ensembl-uniprot map\n",
    "    ensembl_map = pd.read_csv(f'./data/ensembl/{sp}.tsv.gz', sep='\\t')\n",
    "    ensembl_map = ensembl_map[['gene_stable_id', 'xref']].rename(columns={'xref': 'UniProtKB', 'gene_stable_id': 'ensembl_xref'})\n",
    "    zdf = pd.merge(zdf, ensembl_map, how='left', on='UniProtKB')\n",
    "\n",
    "    # join with the uniprot-bgee map\n",
    "    zdf = pd.merge(zdf, bgee_map, how='left', on='UniProtKB')\n",
    "    # these are the genes that we have data on...\n",
    "    expr_df = pd.read_hdf('./results/expr/animal_tpm_expr.h5', sp)\n",
    "    genes_with_data = set(expr_df.index)\n",
    "\n",
    "    # concatenate the different mappings\n",
    "    a = zdf[['UniProtKB', 'Bgee']].rename(columns={'Bgee': 'xref'})\n",
    "    a['type'] = 'uniprot-bgee'\n",
    "    b = zdf[['UniProtKB', 'ensembl_xref']].rename(columns={'ensembl_xref': 'xref'})\n",
    "    b['type'] = 'ensembl-uniprot'\n",
    "    c = zdf[['UniProtKB', 'Ensembl']].rename(columns={'Ensembl': 'xref'})\n",
    "    c['type'] = 'panther_ensembl'\n",
    "    d = zdf[['UniProtKB', 'EnsemblGenome']].rename(columns={'EnsemblGenome': 'xref'})\n",
    "    d['type'] = 'panther_ensemblgenome'\n",
    "\n",
    "    if sp == 'HUMAN':\n",
    "        zdf = pd.merge(zdf, hgnc_map, how='left', left_on='HGNC', right_on='hgnc_id')\n",
    "        #map_df = pd.concat((map_df, zdf[['UniProtKB', 'ensembl_hgnc_map']].rename(columns={'ensembl_hgnc_map': 'xref'})))\n",
    "        e = zdf[['UniProtKB', 'ensembl_hgnc_map']].rename(columns={'ensembl_hgnc_map': 'xref'})\n",
    "        e['type'] = 'hgnc'\n",
    "        map_df = pd.concat((e,d,c,b,a))\n",
    "    elif sp == 'RAT':\n",
    "        zdf = pd.merge(zdf, rgd_map, how='left', left_on='RGD', right_on='rgd_id')\n",
    "        #map_df = pd.concat((map_df, zdf[['UniProtKB', 'ensembl_rgd_map']].rename(columns={'ensembl_rgd_map': 'xref'})))\n",
    "        e = zdf[['UniProtKB', 'ensembl_rgd_map']].rename(columns={'ensembl_rgd_map': 'xref'})\n",
    "        e['type'] = 'rgd'\n",
    "        map_df = pd.concat((e,d,c,b,a))\n",
    "    elif sp == 'MOUSE':\n",
    "        zdf = pd.merge(zdf, mgi_map, how='left', left_on='MGI', right_on='mgi_id')\n",
    "        #map_df = pd.concat((map_df, zdf[['UniProtKB', 'ensembl_mgi_map']].rename(columns={'ensembl_mgi_map': 'xref'})))\n",
    "        e = zdf[['UniProtKB', 'ensembl_mgi_map']].rename(columns={'ensembl_mgi_map': 'xref'})\n",
    "        e['type'] = 'mgi'\n",
    "        map_df = pd.concat((e,d,c,b,a))\n",
    "    else:\n",
    "        # Note: concatenation is ordered and then duplicates removed to prioritise higher quality mappings\n",
    "        # i.e., those in source (e.g., PANTHER or via MGI/RGD), then those from ensembl, then those from bgee.\n",
    "        map_df = pd.concat((d,c,b,a))\n",
    "    map_df = map_df[map_df.xref.notna()]\n",
    "\n",
    "    # filter the map to xref that are in the expression data\n",
    "    map_df = map_df[map_df['xref'].isin(genes_with_data)].drop_duplicates('UniProtKB')\n",
    "\n",
    "    all_panther_fam_genes = set(zdf['UniProtKB'])\n",
    "    have = set(map_df[map_df.UniProtKB.isin(all_panther_fam_genes)].UniProtKB)\n",
    "    missing = all_panther_fam_genes - set(map_df.UniProtKB)\n",
    "    stats.append({'species': sp,\n",
    "                  'dataset': 'animal',\n",
    "                  'n_mapped_with_data': len(have),\n",
    "                  'n_in_fam': len(all_panther_fam_genes)\n",
    "                 })\n",
    "    stats[-1]['percent_with_data'] = '{:02%}'.format(stats[-1]['n_mapped_with_data'] / stats[-1]['n_in_fam'])\n",
    "\n",
    "    res[sp] = map_df[map_df.UniProtKB.isin(all_panther_fam_genes)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82fb0b1f-c488-4b72-9677-597dad04b636",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e63313d3-0e24-4e2b-a18e-e614ed8245d3",
   "metadata": {},
   "source": [
    "## Plant Mapping\n",
    "Check the consistency for the plant mapping files and reformat."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "851672f9-4eeb-4377-83b5-0d8b57f95af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_uniprot(x):\n",
    "    for y in x.split('|'):\n",
    "        if y.startswith('UniProtKB'):\n",
    "            return y.split('=')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b6fe0762-bc91-41db-b902-1e84f6f54d7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pthr_genes_plants = pthr_genes[pthr_genes.species.isin(dataset['plants'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "642d76e4-ff01-4693-a163-71f4e21acea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pthr_genes_plants = {sp: zdf for (sp, zdf) in pthr_genes_plants.groupby('species')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a37b2edd-51aa-4184-9693-ae791af8cf87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e4a4ca7623884d7bb4bc5e074f222e81",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/20 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plant_expr = 'results/tpm_expr/plants'\n",
    "\n",
    "for sp in tqdm(dataset['plants']):\n",
    "    # load the expression to get all gene identifiers\n",
    "    general_df = pd.read_hdf('./results/expr/plant_general_tpm_expr.h5', sp)\n",
    "    specific_df = pd.read_hdf('./results/expr/plant_specific_tpm_expr.h5', sp)\n",
    "    sp_gene_ids = (set(general_df.index) | set(specific_df.index))\n",
    "\n",
    "    # load mapping file\n",
    "    map_df = pd.read_csv(f'data/plant_expr/mapping/{sp}.conversion.panther.txt', sep='\\t', names=['gene_id', 'panther_id'])\n",
    "    map_df['uniprot_id'] = map_df.panther_id.apply(find_uniprot)\n",
    "    assert map_df.uniprot_id.isna().sum() == 0\n",
    "\n",
    "    map_df1 = map_df[map_df.gene_id.isin(sp_gene_ids)]\n",
    "\n",
    "    # identify proportion that we have mapped\n",
    "    stats.append({'species': sp,\n",
    "                  'dataset': 'plant',\n",
    "                  'n_mapped_with_data': len(set(pthr_genes_plants[sp].UniProtKB) & set(map_df1.uniprot_id)),\n",
    "                  'n_in_fam': len(set(pthr_genes_plants[sp].UniProtKB))\n",
    "                 })\n",
    "    stats[-1]['percent_with_data'] = '{:.02%}'.format(stats[-1]['n_mapped_with_data']/ stats[-1]['n_in_fam'])\n",
    "\n",
    "    # save similar to the plants\n",
    "    map_df1 = map_df1[['uniprot_id', 'gene_id']].rename(columns={'uniprot_id': 'UniProtKB', 'gene_id': 'xref'})\n",
    "    map_df1['type'] = 'plant_mapping'\n",
    "    \n",
    "    res[sp] = map_df1[map_df1.UniProtKB.isin(set(pthr_genes_plants[sp].UniProtKB))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4ee1233e-5f49-42fc-b795-739aba31f093",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4ba43d2a63d24b69b27edaf5e1b78367",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/41 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for sp in tqdm(res):\n",
    "    res[sp].to_hdf('./results/expr/gene_map.h5', sp, complevel=9, complib='zlib')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ba189b89-f003-44eb-8c0e-3682a28cdc5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_df = pd.DataFrame(stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6c399990-f66e-40da-bcfa-042c9eac04f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_df.to_csv('./results/tables/gene_map_stats.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4d28218c-7b19-4aff-97d8-de251608a50f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>dataset</th>\n",
       "      <th>n_mapped_with_data</th>\n",
       "      <th>n_in_fam</th>\n",
       "      <th>percent_with_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ANOCA</td>\n",
       "      <td>animal</td>\n",
       "      <td>12149</td>\n",
       "      <td>18355</td>\n",
       "      <td>66.189049%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BOVIN</td>\n",
       "      <td>animal</td>\n",
       "      <td>19996</td>\n",
       "      <td>22367</td>\n",
       "      <td>89.399562%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAEEL</td>\n",
       "      <td>animal</td>\n",
       "      <td>12191</td>\n",
       "      <td>14390</td>\n",
       "      <td>84.718555%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CANLF</td>\n",
       "      <td>animal</td>\n",
       "      <td>565</td>\n",
       "      <td>19481</td>\n",
       "      <td>2.900262%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHICK</td>\n",
       "      <td>animal</td>\n",
       "      <td>14436</td>\n",
       "      <td>16357</td>\n",
       "      <td>88.255793%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DANRE</td>\n",
       "      <td>animal</td>\n",
       "      <td>6946</td>\n",
       "      <td>24063</td>\n",
       "      <td>28.865894%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DROME</td>\n",
       "      <td>animal</td>\n",
       "      <td>10804</td>\n",
       "      <td>10985</td>\n",
       "      <td>98.352299%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FELCA</td>\n",
       "      <td>animal</td>\n",
       "      <td>16747</td>\n",
       "      <td>18549</td>\n",
       "      <td>90.285191%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GORGO</td>\n",
       "      <td>animal</td>\n",
       "      <td>16535</td>\n",
       "      <td>19527</td>\n",
       "      <td>84.677626%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HORSE</td>\n",
       "      <td>animal</td>\n",
       "      <td>17963</td>\n",
       "      <td>19857</td>\n",
       "      <td>90.461802%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HUMAN</td>\n",
       "      <td>animal</td>\n",
       "      <td>18759</td>\n",
       "      <td>19446</td>\n",
       "      <td>96.467140%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>LEPOC</td>\n",
       "      <td>animal</td>\n",
       "      <td>15982</td>\n",
       "      <td>16813</td>\n",
       "      <td>95.057396%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MACMU</td>\n",
       "      <td>animal</td>\n",
       "      <td>19198</td>\n",
       "      <td>20597</td>\n",
       "      <td>93.207749%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MONDO</td>\n",
       "      <td>animal</td>\n",
       "      <td>17842</td>\n",
       "      <td>19448</td>\n",
       "      <td>91.742081%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MOUSE</td>\n",
       "      <td>animal</td>\n",
       "      <td>19303</td>\n",
       "      <td>21081</td>\n",
       "      <td>91.565865%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ORNAN</td>\n",
       "      <td>animal</td>\n",
       "      <td>14898</td>\n",
       "      <td>16217</td>\n",
       "      <td>91.866560%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ORYLA</td>\n",
       "      <td>animal</td>\n",
       "      <td>21065</td>\n",
       "      <td>21482</td>\n",
       "      <td>98.058840%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PANTR</td>\n",
       "      <td>animal</td>\n",
       "      <td>18956</td>\n",
       "      <td>20911</td>\n",
       "      <td>90.650854%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PIG</td>\n",
       "      <td>animal</td>\n",
       "      <td>17553</td>\n",
       "      <td>20554</td>\n",
       "      <td>85.399436%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RAT</td>\n",
       "      <td>animal</td>\n",
       "      <td>17512</td>\n",
       "      <td>21042</td>\n",
       "      <td>83.224028%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>XENTR</td>\n",
       "      <td>animal</td>\n",
       "      <td>9860</td>\n",
       "      <td>18894</td>\n",
       "      <td>52.185879%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SOLTU</td>\n",
       "      <td>plant</td>\n",
       "      <td>22240</td>\n",
       "      <td>24371</td>\n",
       "      <td>91.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SOYBN</td>\n",
       "      <td>plant</td>\n",
       "      <td>43975</td>\n",
       "      <td>46939</td>\n",
       "      <td>93.69%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ORYSJ</td>\n",
       "      <td>plant</td>\n",
       "      <td>22200</td>\n",
       "      <td>24745</td>\n",
       "      <td>89.72%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CUCSA</td>\n",
       "      <td>plant</td>\n",
       "      <td>17878</td>\n",
       "      <td>18186</td>\n",
       "      <td>98.31%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PHYPA</td>\n",
       "      <td>plant</td>\n",
       "      <td>18454</td>\n",
       "      <td>19248</td>\n",
       "      <td>95.87%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MEDTR</td>\n",
       "      <td>plant</td>\n",
       "      <td>27439</td>\n",
       "      <td>31996</td>\n",
       "      <td>85.76%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ARATH</td>\n",
       "      <td>plant</td>\n",
       "      <td>23809</td>\n",
       "      <td>24292</td>\n",
       "      <td>98.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>AMBTC</td>\n",
       "      <td>plant</td>\n",
       "      <td>14765</td>\n",
       "      <td>15625</td>\n",
       "      <td>94.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>BRANA</td>\n",
       "      <td>plant</td>\n",
       "      <td>39226</td>\n",
       "      <td>41424</td>\n",
       "      <td>94.69%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HELAN</td>\n",
       "      <td>plant</td>\n",
       "      <td>28694</td>\n",
       "      <td>36296</td>\n",
       "      <td>79.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>VITVI</td>\n",
       "      <td>plant</td>\n",
       "      <td>16777</td>\n",
       "      <td>21625</td>\n",
       "      <td>77.58%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CAPAN</td>\n",
       "      <td>plant</td>\n",
       "      <td>23173</td>\n",
       "      <td>26381</td>\n",
       "      <td>87.84%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MARPO</td>\n",
       "      <td>plant</td>\n",
       "      <td>10640</td>\n",
       "      <td>11375</td>\n",
       "      <td>93.54%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SETIT</td>\n",
       "      <td>plant</td>\n",
       "      <td>22637</td>\n",
       "      <td>26013</td>\n",
       "      <td>87.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SOLLC</td>\n",
       "      <td>plant</td>\n",
       "      <td>22416</td>\n",
       "      <td>23974</td>\n",
       "      <td>93.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>WHEAT</td>\n",
       "      <td>plant</td>\n",
       "      <td>85467</td>\n",
       "      <td>92511</td>\n",
       "      <td>92.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MAIZE</td>\n",
       "      <td>plant</td>\n",
       "      <td>27598</td>\n",
       "      <td>28210</td>\n",
       "      <td>97.83%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>POPTR</td>\n",
       "      <td>plant</td>\n",
       "      <td>27108</td>\n",
       "      <td>32654</td>\n",
       "      <td>83.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>SELML</td>\n",
       "      <td>plant</td>\n",
       "      <td>23665</td>\n",
       "      <td>25897</td>\n",
       "      <td>91.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>MANES</td>\n",
       "      <td>plant</td>\n",
       "      <td>25101</td>\n",
       "      <td>26416</td>\n",
       "      <td>95.02%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species dataset  n_mapped_with_data  n_in_fam percent_with_data\n",
       "0    ANOCA  animal               12149     18355        66.189049%\n",
       "1    BOVIN  animal               19996     22367        89.399562%\n",
       "2    CAEEL  animal               12191     14390        84.718555%\n",
       "3    CANLF  animal                 565     19481         2.900262%\n",
       "4    CHICK  animal               14436     16357        88.255793%\n",
       "5    DANRE  animal                6946     24063        28.865894%\n",
       "6    DROME  animal               10804     10985        98.352299%\n",
       "7    FELCA  animal               16747     18549        90.285191%\n",
       "8    GORGO  animal               16535     19527        84.677626%\n",
       "9    HORSE  animal               17963     19857        90.461802%\n",
       "10   HUMAN  animal               18759     19446        96.467140%\n",
       "11   LEPOC  animal               15982     16813        95.057396%\n",
       "12   MACMU  animal               19198     20597        93.207749%\n",
       "13   MONDO  animal               17842     19448        91.742081%\n",
       "14   MOUSE  animal               19303     21081        91.565865%\n",
       "15   ORNAN  animal               14898     16217        91.866560%\n",
       "16   ORYLA  animal               21065     21482        98.058840%\n",
       "17   PANTR  animal               18956     20911        90.650854%\n",
       "18     PIG  animal               17553     20554        85.399436%\n",
       "19     RAT  animal               17512     21042        83.224028%\n",
       "20   XENTR  animal                9860     18894        52.185879%\n",
       "21   SOLTU   plant               22240     24371            91.26%\n",
       "22   SOYBN   plant               43975     46939            93.69%\n",
       "23   ORYSJ   plant               22200     24745            89.72%\n",
       "24   CUCSA   plant               17878     18186            98.31%\n",
       "25   PHYPA   plant               18454     19248            95.87%\n",
       "26   MEDTR   plant               27439     31996            85.76%\n",
       "27   ARATH   plant               23809     24292            98.01%\n",
       "28   AMBTC   plant               14765     15625            94.50%\n",
       "29   BRANA   plant               39226     41424            94.69%\n",
       "30   HELAN   plant               28694     36296            79.06%\n",
       "31   VITVI   plant               16777     21625            77.58%\n",
       "32   CAPAN   plant               23173     26381            87.84%\n",
       "33   MARPO   plant               10640     11375            93.54%\n",
       "34   SETIT   plant               22637     26013            87.02%\n",
       "35   SOLLC   plant               22416     23974            93.50%\n",
       "36   WHEAT   plant               85467     92511            92.39%\n",
       "37   MAIZE   plant               27598     28210            97.83%\n",
       "38   POPTR   plant               27108     32654            83.02%\n",
       "39   SELML   plant               23665     25897            91.38%\n",
       "40   MANES   plant               25101     26416            95.02%"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f131914-8a8b-4aa5-a270-2da27cfc1dc5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
